language/move-prover/bytecode/src/verification_analysis_v2.rs

```diff
@@ -169,37 +170,87 @@ fn compute_disabled_invs_for_fun(
 /// The second condition is an exception for functions that cannot invalidate
 /// any invariants.
 fn check_legal_disabled_invariants(
-    _fun_env: &FunctionEnv,
-    _disabled_inv_fun_set: &BTreeSet<QualifiedId<FunId>>,
-    _non_inv_fun_set: &BTreeSet<QualifiedId<FunId>>,
-    _funs_that_modify_some_inv: &BTreeSet<QualifiedId<FunId>>,
+    fun_env: &FunctionEnv,
+    disabled_inv_fun_set: &BTreeSet<QualifiedId<FunId>>,
+    non_inv_fun_set: &BTreeSet<QualifiedId<FunId>>,
+    funs_that_modify_some_inv: &BTreeSet<QualifiedId<FunId>>,
 ) {
-
-    // TODO: this currently produces a false positive for unknown reasons. It is hard to
-    //   debug because the error message does not give any info about the source of
-    //   invariant disabling. To give better error messages, we need to extend the analysis
-    //   to track locations.
-    /*
     let global_env = fun_env.module_env.env;
     let fun_id = fun_env.get_qualified_id();
-    if non_inv_fun_set.contains(&fun_id) && funs_that_modify_some_inv.contains(&fun_id) {
+    let explicitly_disabled =
+        |fenv: &FunctionEnv<'_>| disabled_inv_fun_set.contains(&fenv.get_qualified_id());
+    let implicitly_disabled = |fenv: &FunctionEnv| {
+        non_inv_fun_set.contains(&fun_env.get_qualified_id())
+            && funs_that_modify_some_inv.contains(&fenv.get_qualified_id())
+    };
+    if fun_env.has_unknown_callers() {
         // function is public or script, which means we don't know whether invariants hold
-        // or not at the call site
-        if fun_env.has_unknown_callers() {
+        if explicitly_disabled(fun_env) {
             global_env.error(
                 &fun_env.get_loc(),
-                "Public or script functions must not be called when invariants are disabled \
-                 in a transitive caller or there is a pragma delegate_invariants_to_caller",
-            );
-        } else if disabled_inv_fun_set.contains(&fun_id) {
+                "Public or script functions cannot disable invariants",
```

Comment:
This design is problematic and not the expected one in the context of the generic invariant work. With that design we can't deal with the generic invariant problem Junkil reported. The generic invariant work started under the assumption that invariants disabled in a function are evaluated at the caller side, not at the return of the called function. This is necessary so we can evaluate the invariant with the type instantiation at the caller side and in the context of assumptions at the caller side.

Interestingly enough, you are already evaluating at caller side for opaque functions (though there is a bug because you don't treat update invariants in opaque calls, related to Junkil's issue). So currently there is a discrepancy between how opaque and inlined functions are dealt with.

I like to suggest we switch to evaluating all disabled invariants at caller side. This is not only the only way I know how to deal with generic invariants, but also appears more logical to me in general, as the assumptions in caller context can be used to prove the invariants.

Reply:
Question: if we use delegate_invariants_to_caller explicitly, would then the invariants be evaluated at the caller side, and we have the effect we need for generic functions?

If that is the case, perhaps we should only support the delegate feature. I find it very difficult to distinguish the subtleties of those both things (and the implementation of both features at the same time is rather complex).